presto-spi/src/main/java/io/prestosql/spi/type/AbstractType.java

```diff
@@ -16,19 +16,49 @@
 import io.airlift.slice.Slice;
 import io.prestosql.spi.block.Block;
 import io.prestosql.spi.block.BlockBuilder;
+import io.prestosql.spi.function.InvocationConvention;
+import io.prestosql.spi.function.OperatorInvoker;
+import io.prestosql.spi.function.ScalarFunctionAdapterCache;
 
+import java.util.Arrays;
+import java.util.Collections;
 import java.util.List;
 
+import static java.util.Collections.emptyList;
+import static java.util.Objects.requireNonNull;
+
 public abstract class AbstractType
         implements Type
 {
     private final TypeSignature signature;
     private final Class<?> javaType;
+    private final ScalarFunctionAdapterCache equalsOperatorCache;
+    private final ScalarFunctionAdapterCache hashCodeOperatorCache;
```

Comment:
This is how this could be modeled. Basically, let the `getXXXOperator()` methods take a context object they can use to get the invokers associated with their type dependencies, if any. That context would be in charge of maintaining the cache and dispatching to the element type if necessary:

![getOperator](https://user-images.githubusercontent.com/14387/85080997-995a0d00-b17f-11ea-81d7-ef02ce006982.jpg)